### ISSUE TITLE
Is negative zero a default value for float/double fields?

### ISSUE THREAD

**Issue by tbkka** [protocolbuffers/protobuf#7062]
This question recently came up in discussion of apple/swift-protobuf#935:

> Should proto3 implementations serialize float/double values of `-0.0`?

Proto3 omits any field whose value is the default.  For float and double fields, this means that zero values are not serialized.  Does this apply to both negative and positive zero?  Or should negative zero be serialized in order to ensure that all float and double values can be accurately communicated?

**Comment by tbkka** [comment]
The desired behavior should probably be included in conformance tests for binary, text format, and JSON serializations.

**Comment by acozzette** [comment]
CC @gerben-s who was looking into this problem a couple months ago

**Comment by elharo** [comment]
While we're at it, I wonder what we do with NaN and Inf?

**Comment by fmeum** [comment]
We ran into this issue while working on a Java-based mutator for Protobuf messages and have to rely on [rather hacky workarounds](https://github.com/CodeIntelligenceTesting/jazzer/blob/7e234b6e73200253eaf2069ad0ad57b7db4e212d/src/test/java/com/code_intelligence/jazzer/mutation/mutator/StressTest.java#L518-L562). Is there a plan to settle this edge case?

**Comment by tbkka** [comment]
> While we're at it, I wonder what we do with NaN and Inf?

NaN and Inf are transferred correctly in all cases already.  There's no problem there.  The problem with `-0.0` is that it tests equal to `+0.0`.  So the question here is really whether a field that has the value `-0.0`:
* **Should** be transferred because it is not the same as the default `+0.0`
* **Should not** be transferred because it is equal to the default `+0.0`

Or put different, is the default `+0.0` or is it "any double value that is equal to `+0.0`"?

**Comment by fmeum** [comment]
While `0.0 == -0.0` in Java, `Double.compare(0.0, -0.0)` returns `1`, not `0`, so this even depends on the kind of equality test.

**Comment by fmeum** [comment]
There is an important case in which `-0.0` and `0.0` are _not_ treated as equivalent: When comparing proto message objects, at least in Java. As it stands, two proto message objects with `float32` proto fields, one set to `0.0` and one set to `-0.0`, will not test equal in Java, but will do so after having been written to bytes and parsed back. 

Is this still expected behavior? I would have expected `equals` semantics in Java to be identical to be preserved under a serialization/deserialization roundtrip.

**Comment by fowles** [comment]
I would have expected that as well.  Honestly, this particular corner of our semantics is likely under covered in tests

**Comment by bilbothebaggins** [comment]
Let me add a quick quote from the specs here: https://protobuf.dev/programming-guides/encoding/

> ```
> i64        := sfixed64 | fixed64 | double;
>                 encoded as 8-byte little-endian;
>                 memcpy of the equivalent C types (u?int64_t, double)
> ```

If we take the spec at face value **memcpy of the equivalent C types ... double)** than _all_ the behaviors that do not preserve the bit pattern are "wrong".

**Comment by mumbleskates** [comment]
As of [2024-02-13](https://github.com/protocolbuffers/protocolbuffers.github.io/commit/025f35021d3d1f3a4d87e150f6727511977f8c6a), this was changed:

> Also note that if a scalar message field **is** set to its default, the value will not be serialized on the wire. If a float or double value is set to +0 it will not be serialized, but -0 is considered distinct and will be serialized.

I cannot tell if or when this has become true in which official protobuf libraries, or if or when this has existed in any part of the conformance tests. There have been [at least some unit tests](https://github.com/protocolbuffers/protobuf/blame/d1cf09abcff7636101fd27e94ea5aa7b9f603f5e/java/core/src/test/java/com/google/protobuf/MessageTest.java#L362-L372) since 2021. it also seems to work in c++ at present...

**Comment by bilbothebaggins** [comment]
This issue should be closed as "answered"?

There is the commit https://github.com/protocolbuffers/protocolbuffers.github.io/commit/025f35021d3d1f3a4d87e150f6727511977f8c6a from that changed the [current docs](https://protobuf.dev/programming-guides/proto3/#default) to the quote already mentioned above:

> ... for *scalar* message fields, once a message is parsed there’s no way of telling whether a field was explicitly set to the default value ...  If a float or double value is set to `+0` it will not be serialized, but `-0` is considered distinct and will be serialized.

**Comment by tbkka** [comment]
Ummm....  so which is it?

The [current docs](https://protobuf.dev/programming-guides/proto3/#default) say:
>  If a float or double value is set to +0 it will not be serialized, but -0 is considered distinct and will be serialized.

but

@Logofile just above claimed 
> ... "[i]f a float or double value is set to -0 or +0, it will not be serialized"

**Comment by Logofile** [comment]
Apologies: that additional text was removed in a later change. (I was looking at an earlier version of the `proto3.md` file when I added my last comment.) We removed the text "For float and double types, -0.0 and 0.0 are treated as equivalent, and will round-trip" based on guidance from @esrauchg.  

If a float or double value is set to +0 it **will not** be serialized, but -0 **will** be serialized.
